src/noises.jl

```diff
@@ -24,19 +24,147 @@ struct NetworkPauliNoise
+                # for every qubit that is not up to speed with the max time
+                for q in acted_on_qubits
+                    if accounted_for_time[q] < max_time
+                        # this qubit has not been accounted for up until now, so we must add T1 and T2 noise here now
+                        
+                        # get the error operations from time idling, t1, t2 values, and push them
+                        idle_time = max_time - accounted_for_time[q]
+                        λ₁, λ₂ = thermal_relaxation_error_rate(error.t1, error.t2, idle_time)
+                        push!(noisy_circuit, T1NoiseOp(q,λ₁), T2NoiseOp(q,λ₂))
+
+                        # now update its time
+                        accounted_for_time[q] = max_time # set it to the maximum time, so that it is now in sync with the other qubits
+                    end
+                end
+            end
+
+            # now all important qubits are up to speed, we can apply the operation 
+            push!(noisy_circuit, op)
+
+            # and update the time elapsed for this operation 
+            for q in acted_on_qubits
+                accounted_for_time[q] += time_to_elapse # update the time for all qubits that were acted on
+            end
+
+            ## --- After applying operation ---
+        end
+    end
+
+    # Now all operations have T1T2 noise padded before, and in-between them if needed. So now we need to bring all of the qubits up to speed with the end time.
+    # Get the max time on a qubit (this is the time that it takes to execute the entire circuit)
+    max_time = maximum(accounted_for_time)
+
+    # Bring all qubits up to speed with the maximum time (pad with T1T2)
+    for q in qubit_list
+        if accounted_for_time[q] < max_time  
+            # this qubit has not been accounted for up until now, so we must add T1 and T2 noise here now
+            
+            # get the error operations from time idling, t1, t2 values, and push them
+            idle_time = max_time - accounted_for_time[q]
+            λ₁, λ₂ = thermal_relaxation_error_rate(error.t1, error.t2, idle_time)
+            push!(noisy_circuit, T1NoiseOp(q,λ₁), T2NoiseOp(q,λ₂))
+        end
+    end
+
+    # TODO: make sure we can simplify/compactify_circuit with QuantumClifford
+
+    return noisy_circuit
+end
+    
+# thermal relaxation error from t1 and t2 values, and the time elapsed. From YipiaoWu/QuantumHardware
+function thermal_relaxation_error_rate(t1, t2, gate_time) 
+    λ₁ = 1 - exp(-gate_time/t1)
+    t_ϕ = t1*t2 / (2*t1 - t2)
+    λ₂ = 1 - exp(-gate_time/t_ϕ)
+    return λ₁, λ₂
+end
+
+# retrieve qubits involved in a gate
+affectedqubits(gate::PauliNoiseBellGate) = (gate.g.idx1, gate.g.idx2)
+affectedqubits(gate::NoisyBellMeasureNoisyReset) = (gate.m.sidx,)
+affectedqubits(gate::BellMeasure) = (gate.sidx,)
+affectedqubits(gate::CNOTPerm) = (gate.idx1, gate.idx2)
+affectedqubits(gate)= ()  # Default case for gates that do not involve qubits
```

Comment:
in the call we discussed the tradeoffs between naming clarity and code simplicity about using the names `affectedqubits` vs `affectedpairs`.

If you decide to call it `affectedqubits` could you do these as methods that are defined in BPGates.jl.

If you decide to call it `affectedpairs` could you define such a function in BPGates.jl and put these as method definitions in BPGates.jl